File: retreat-be/retreat-be/src/main/java/main/retreats/models/client/Event.java
package main.retreats.models.client;

import jakarta.persistence.*;
import jakarta.validation.constraints.NotNull;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import main.retreats.models.enums.EventType;
import main.retreats.models.enums.RequestStatus;

import java.time.LocalDate;
import java.util.HashSet;
import java.util.Set;

@Getter
@Setter
@NoArgsConstructor
@Entity
public class Event {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @NotNull
    @Column(nullable = false)
    private String name;

    @NotNull
    @Column(nullable = false)
    private String description;

    @NotNull
    @Column(nullable = false)
    private LocalDate createdOn;

    @NotNull
    @Column(nullable = false)
    private LocalDate startTime;

    @NotNull
    @Column(nullable = false)
    private LocalDate endTime;

    @NotNull
    @OneToMany(cascade = CascadeType.ALL, orphanRemoval = true)
    @JoinColumn(name = "fk_event", referencedColumnName = "id")
    private Set<EventPrice> prices = new HashSet<>();

    @NotNull
    @Column(nullable = false)
    @Enumerated(EnumType.STRING)
    private EventType type;

    @ManyToMany(cascade = CascadeType.DETACH)
    @JoinTable(name = "event_event_category",
                joinColumns = {@JoinColumn(name="fk_event", referencedColumnName = "id")},
                inverseJoinColumns = {@JoinColumn(name="fk_category", referencedColumnName = "id")})
    private Set<EventCategory> category;

    private Integer maxParticipants;

    private String location;

    private String locationGooglePin;

    @ManyToOne
    @JoinColumn(name = "fk_cluster")
    private EventCluster cluster;

    private String additionalInformation;

    @NotNull
    @Column(nullable = false)
    @Enumerated(EnumType.STRING)
    private RequestStatus status;

    @NotNull
    @Column(nullable = false)
    private LocalDate statusDate;
}

File: retreat-fe/src/app/services/models/event.ts
export class Event {
  constructor(public id: number | null,
              public name: string,
              public description: string,
              public startTime: Date,
              public endTime: Date,
              public prices: EventPrice[],
              public type: EventType,
              public category: EventCategory,
              public maxParticipants: number | null,
              public location: string | null,
              public locationGooglePin: string | null,
              public additionalInformation: string | null,
              public status: string = 'REQUESTED'){ }
}

export class EventPrice {
  constructor(public id: number | null,
              public fromDate: Date | null,
              public toDate: Date | null,
              public price: number,
              public description: string | null) { }
}

export enum EventType {
  RETREAT = "RETREAT",
  EVENT = "EVENT"
}

export class EventCategory {
  constructor(public id: number | null,
              public name: string,
              public description: string | null,
              public status: string = 'REQUESTED',
              public parentId: number | null,
              public imageUrl: string | null,
              public requestedBy: string) { }
}
export enum CurrencyType {
  BGN = "BGN",
  EUR = "EUR"
}

export const activeCurrencies: CurrencyType[] = [
  CurrencyType.BGN
]

export enum REPEAT_INTERVAL {
  MONTHLY = "MONTHLY",
  WEEKLY = "WEEKLY",
  EVERY_YEAR = "EVERY_YEAR",
  CUSTOM_INTERVAL = "CUSTOM_INTERVAL"
}
